go-controller/pkg/node/ovn_db_mgmt.go

```diff
@@ -0,0 +1,139 @@
+package node
+
+import (
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+func (n *OvnNode) ensureOvnRaft(db string) {
+	ticker := time.NewTicker(60 * time.Second)
+	klog.Infof("Starting ensure routine for Raft db: %s", db)
+	kclient := n.Kube.(*kube.Kube)
+	for {
+		select {
+		case <-ticker.C:
+			if dbOnNode, _ := metrics.CheckPodRunsOnGivenNode(kclient.KClient.(*kubernetes.Clientset),
+				"name=ovnkube-db", n.name, false); dbOnNode {
+				klog.Infof("Node %s has a OVN DB pod, ensuring OVN Raft membership", n.name)
+				n.ensureLocalRaftServerID(db)
+				n.ensureClusterRaftMembership(db)
+			}
+		case <-n.stopChan:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// ensureLocalRaftServerID is used to ensure there is no stale member in the Raft cluster with our address
+func (n *OvnNode) ensureLocalRaftServerID(db string) {
+	var dbName string
+	var appCtl func(args ...string) (string, string, error)
+
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+	} else {
+		dbName = "OVN_Southbound"
+		appCtl = util.RunOVNSBAppCtl
+	}
+
+	out, stderr, err := util.RunOVSDBTool("db-sid", db)
+	if err != nil {
+		klog.Warningf("Unable to get db server ID for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	if len(out) < 4 {
+		klog.Errorf("Invalid db id found: %s for db: %s", out, db)
+		return
+	}
+	// server ID in raft membership is only first 4 char prefix
+	serverID := out[:4]
+	out, stderr, err = appCtl("cluster/status", dbName)
+	if err != nil {
+		klog.Warningf("Unable to get cluster status for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	r, _ := regexp.Compile(`Address: *((ssl|tcp):[?[a-z0-9.:]+]?)`)
+	matches := r.FindStringSubmatch(out)
+	if len(matches) < 2 {
+		klog.Warningf("Unable to parse Address for db: %s, output: %s", db, out)
+		return
+	}
+	addr := matches[1]
+	// look for current servers in raft cluster with the same address
+	r, _ = regexp.Compile("([a-z0-9]{4}) at " + addr)
+	members := r.FindAllStringSubmatch(out, -1)
+	for _, member := range members {
+		if len(member) < 2 {
```

Comment:
the `len(members)` will be 1 since your regex is bound to `addr`? 

Looks like some of the code is from when it was running from the `ovnkube-master`?

Reply:
its a submatch so there should be 2 entries, the entire match, plus the submatch of the ID.

Which code are you talking about from ovnkube-master?

Reply:
@trozet I am talking about `members` and not `member`. The `members` length will be always one since you are matching the output against the local node's address, right?

Reply:
@girishmg no that's not the case. The issue this PR is solving is when you bring back a node with a new db it will actually have 2 entries in the raft cluster with the same address: one for the previous ID and one for the new ID:

```
[root@ovn-control-plane2 ~]# ovn-appctl -t /var/run/ovn/ovnnb_db.ctl cluster/status OVN_Northbound
e60c
Name: OVN_Northbound
Cluster ID: bcb7 (bcb74ae8-ba91-47fa-afe8-f436829b079e)
Server ID: e60c (e60cff74-fa76-4bb4-96d6-4c740dd615f9)
Address: tcp:[172.18.0.5]:6643
Status: cluster member
Role: leader
Term: 1
Leader: self
Vote: self

Election timer: 1000
Log: [2, 185]
Entries not yet committed: 0
Entries not yet applied: 0
Connections: <-9d15 ->d265 ->9d15 <-13f8
Servers:
    9d15 (9d15 at tcp:[172.18.0.3]:6643) next_index=185 match_index=184
    e60c (e60c at tcp:[172.18.0.5]:6643) (self) next_index=2 match_index=184
    13f8 (13f8 at tcp:[172.18.0.4]:6643) next_index=185 match_index=184
    d265 (d265 at tcp:[172.18.0.4]:6643) next_index=185 match_index=182
```